### PR TITLE
Optional dtype for reduce functions

### LINKED ISSUE #6593: [JIT] Handle dtypes
Over in https://github.com/pytorch/pytorch/pull/6573 I added support for optional dtypes via specifying an `optional<ScalarType>=nullopt` parameter, but this fails in the jit because the jit doesn't know how to handle ScalarTypes and/or optional values.

For now I believe I can break these up into two separate functions, with a couple of downsides:
1) Can't trace the dtype version
2) The python arg parser will generate two different overloads (which leads to slower parsing and ugly error messages) even though there is really only a single function.

CC @zdevito @apaszke.

### PR DESCRIPTION
**Pull request by tugrulates**
Fixes #6593

This replaces the separately defined overloads for reduce functions that take `dtype` to take this param by `ScalarType?`, now that it is supported in schema. This simplifies much of the logic to support these overloads and introduces the pattern to take `dtype` by kwargs for future needs.

This PR has some minor backward breaking changes:
- dtype by arg is being deleted for `torch.softmax` and `torch.log_softmax`, i.e. `torch.softmax(x, 0, torch.float)` will fail and need to be replaced by `torch.softmax(x, 0, dtype=torch.float)`.
- Calls within script to these functions without dtype (e.g. `sum(%x)`) will not work across versions because kwargs are lost in JIT. See changed expect files for graph changes.
- Some autograd graphs will change from one depth to two depths, due to type cast in the backward  functions. This will break if `retain_graph=False` and multiple backwards are computed on the same path. See the changed test for a real-life example of this.

It might be easier to review this commit-by-commit.

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -331,7 +331,8 @@ def group(l, group_size):
         # Compute the d loss / d intermediates in chunks of shard_size
         shard_size = 2
         d_intermediates = [d_i for intermediates_batch in group(intermediates, shard_size)
-                           for d_i in torch.autograd.grad(loss, intermediates_batch)]
+                           for d_i in torch.autograd.grad(loss, intermediates_batch,
+                                                          retain_graph=True)]
```
Comment (tugrulates) [inline test/test_autograd.py]: This is because the graph has one more depth due to type cast.

**Comment by tugrulates** [comment]
This is now ready to review.

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -597,17 +596,6 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
       });
       return {backward_value->node()->output(0), nullptr, nullptr, nullptr, nullptr};
 
-    } else if (node->matches("aten::log_softmax(Tensor self, int dim) -> Tensor")) {
```
Comment (tugrulates) [inline torch/csrc/jit/autodiff.cpp]: This is already working fine through dispatch and derivatives.yaml. Rather than updating here, simply dropping the override.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1477,11 +1477,7 @@
 - func: smm(Tensor self, Tensor mat2) -> Tensor
   variants: function, method
 
-# FIXME: These could be combined as optional<ScalarType> but for https://github.com/pytorch/pytorch/issues/6593.
-- func: softmax(Tensor self, int64_t dim, ScalarType dtype) -> Tensor
-  variants: function, method
-
-- func: softmax(Tensor self, int64_t dim) -> Tensor
+- func: softmax(Tensor self, int64_t dim, *, ScalarType? dtype=None) -> Tensor
```
Comment (wanchaol) [inline aten/src/ATen/native/native_functions.yaml]: For `softmax` and `log_softmax`, why do we want to make it a kw only argument instead of simply merging them together? It is indeed a BC change and I don't know if we want to do that or not.

File: torch/csrc/jit/autodiff.cpp (hunk 2 above)
Comment (tugrulates) [reply torch/csrc/jit/autodiff.cpp]: Nm, this seems to be used in symbolic. diff. Will bring back.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -236,21 +186,17 @@ static Tensor& prod_out(Tensor& result, const Tensor& self, IntList dim,
   return result;
 }
 
-static Tensor prod(const Tensor& self, IntList dim, bool keepdim, optional<ScalarType> dtype) {
+Tensor prod(const Tensor &self, optional<ScalarType> dtype) {
   Tensor result;
-  native::prod_out(result, self, dim, keepdim, dtype);
+  return at::native::prod_impl(result, self, {}, false, dtype);
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: you have two returns here?

File: aten/src/ATen/native/ReduceOps.cpp [hunk 5]
```diff
@@ -236,21 +186,17 @@ static Tensor& prod_out(Tensor& result, const Tensor& self, IntList dim,
...
   return result;
 }
 
-Tensor prod(const Tensor &self, ScalarType dtype) {
-  return at::native::prod(self, {}, false, optional<ScalarType>(dtype));
-}
+// \ALL REDUCE ################################################################
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: these comments don't seem relevant anymore.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 6]
```diff
@@ -198,34 +166,16 @@ static ScalarType get_dtype(Tensor& result, const Tensor& self, optional<ScalarT
...
 
-Tensor sum(const Tensor &self) {
-  return at::native::sum(self, {}, false, c10::nullopt);
+Tensor sum(const Tensor &self, optional<ScalarType> dtype) {
+  return at::native::sum(self, {}, false, dtype);
 }
 
-static Tensor& prod_out(Tensor& result, const Tensor& self, IntList dim,
-                        bool keepdim, optional<ScalarType> opt_dtype) {
+static Tensor& prod_impl(Tensor& result, const Tensor& self, IntList dim,
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: I don't really understand the organization here.  It's not grouped by function name, which makes it difficult to figure out how a specific function is implemented.  It's also not really grouped by implementation type, i.e. some functions have fn_impl, some fn_out, etc.  I'd just go back to function name grouping.

Comment (gchanan) [reply aten/src/ATen/native/ReduceOps.cpp]: Why is there a prod_impl but not a sum_impl?  What's the distinction?

File: tools/autograd/gen_python_functions.py [hunk 7]
```diff
@@ -768,7 +768,7 @@ def get_py_formal_arg(arg):
         default = None
         if arg.get('default') is not None:
             default = arg['default']
-            if default == 'nullptr' or default == 'nullopt' or default == '{}':
+            if default == 'nullptr' or default == 'c10::nullopt' or default == '{}':
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: I think I asked this before (but perhaps the comment went away because of the force push) -- why do you need the c10?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (tugrulates) [reply aten/src/ATen/native/native_functions.yaml]: It's mainly for the sake of consistency with other ops.

Some pro arguments are:
- These two are undocumented functions. The public interface is through `torch.nn.functional.[log_]softmax` which already have `dtype` as a kwarg.
- kwarg and non-kwarg forms are the same cpp signature so it's tricky to make them live side-by-side for a deprecation period.
- If this change is ever to be made, it's better to do it sooner than later.
- It's a better baseline ever if more kwargs are added across multiple ops (`casting`, `where` etc.).
- The fix is trivial.

File: test/test_jit.py [hunk 8]
```diff
@@ -8801,6 +8849,17 @@ def fn(x, scale, shift):
         with self.capture_stdout() as captured:
             print(fn(x, scale, shift))
 
+    def test_optional_scalartype(self):
+        @torch.jit.script
+        def fn(a, b):
+            return torch.add(a.cumsum(0, dtype=torch.long).sum(dtype=None),
+                             b.cumprod(0, dtype=None).prod(dtype=torch.double))
+
+        a = torch.randn(4, 4, dtype=torch.float, requires_grad=True)
```
Comment (gchanan) [inline test/test_jit.py]: do a and b do anything here?

File: test/test_jit.py [hunk 9]
```diff
@@ -6251,12 +6251,60 @@ def forward(self, x):
...
+
+        mte = ModuleToExport()
+        outputs = mte(torch.zeros(1, 2, 3))
+        with self.assertRaisesRegex(RuntimeError, "Couldn't export operator"):
+            torch.onnx.export_to_pretty_string(
+                mte, (torch.zeros(1, 2, 3),), None, verbose=False,
+                example_outputs=outputs)
+
+    def test_onnx_export_script_module_unsupported_optional(self):
+        class ModuleToExport(torch.jit.ScriptModule):
```
Comment (gchanan) [inline test/test_jit.py]: how is this different than the test above?

File: test/test_jit.py [hunk 10]
```diff
@@ -6251,12 +6251,60 @@ def forward(self, x):
                     x = torch.neg(x)
                 return x
 
+    def test_onnx_export_script_module_overload(self):
+        class ModuleToExport(torch.jit.ScriptModule):
```
Comment (gchanan) [inline test/test_jit.py]: didn't you clobber the test above?

File: tools/autograd/derivatives.yaml [hunk 11]
```diff
@@ -217,17 +217,11 @@
   self: other.cross(grad, dim)
   other: grad.cross(self, dim)
 
-- name: cumprod(Tensor self, int64_t dim)
-  self: cumprod_backward(grad, self, dim)
+- name: cumprod(Tensor self, int64_t dim, *, ScalarType? dtype)
+  self: cumprod_backward(grad.to(self.type()), self, dim)
```
Comment (gchanan) [inline tools/autograd/derivatives.yaml]: For all these functions:
- does this generate an additional cast if we trace it for the case where a dtype isn't passed?

- Also, can you use scalar_type() instead of type()?

File: tools/autograd/gen_python_functions.py (hunk 7 above)
Comment (tugrulates) [reply tools/autograd/gen_python_functions.py]: Ah, responded on https://github.com/pytorch/pytorch/pull/15154 for this.

native_parse marks the default as c10::nullopt and then that gets propagated everywhere.

Rather than adding a bunch of `using` statements in generated code replacing the previous check since that isn't used for anything. Verified by tracing this line in code gen.

File: test/test_jit.py (hunk 10 above)
Comment (tugrulates) [reply test/test_jit.py]: The above test is verifying the `if` behavior. Did you mean this comment for the test below?

File: test/test_jit.py (hunk 9 above)
Comment (tugrulates) [reply test/test_jit.py]: This one is verifying that `dtype` can't be exported to ONNX, and raises and exception.

The one above is verifying that the overload switch I added to `symbolic.py`, i.e. `torch.prod(x)` matches `prod(x, None)` export but `torch.prod(x, 0)` matches `prod(x, dim, keepdim, None)` export.

File: test/expect/TestBatched.test_while.expect [hunk 12]
```diff
@@ -11,33 +11,34 @@ graph(%a.1_data : Tensor
   %10 : Tensor = aten::__or__(%a.1_dims, %b_dims)
   %11 : int = prim::Constant[value=0]()
   %12 : Tensor = aten::mul(%8, %9)
-  %13 : Tensor = aten::sum(%12)
-  %14 : Tensor = aten::gt(%13, %11)
-  %15 : bool = prim::Bool(%14)
-  %16 : Tensor, %17 : Tensor, %a : Tensor, %19 : Tensor, %20 : Tensor = prim::Loop(%7, %15, %8, %9, %a.1_data, %a.1_mask, %a.1_dims)
+  %13 : int? = prim::None()
```
[the remark this answers is not in the document]
Comment (tugrulates) [reply test/expect/TestBatched.test_while.expect]: This is not calling a cast, but adding a `prim::None` to `sum` symbol in line 15 below.

This is because this PR is removing the overloads without the dtypes, which is kind of the purpose of this PR.

Since the generated graphs include the default values from the schema, this will keep happening with every new kwarg added to ops. Two ways to fix this:
1. Keep overloads in `native_functions.yaml` and keep adding them with every new arg. (makes this PR unneeded)
2. Make JIT apply defaults at execution time. Maybe this behavior might be confined to kwargs only. (much larger scope than what this PR trying to achieve)

File: test/test_autograd.py (hunk 1 above)
Comment (tugrulates) [reply test/test_autograd.py]: The type cast will be no-op when dtype is not passed, but still adds another node to the graph. To avoid it altogether, we'll need to provide an overload to the native function without the `dtype` kwarg.

The behavior that is surprising is that `retain_graph` not being needed for the simplest backwards cases. The documentation says the grad graph will be freed, yet this test still manages to use it more than once. This is not a problem for just this PR. This quirk will apply with any change to any derivative, when the node count goes from one to two.

File: tools/autograd/derivatives.yaml (hunk 11 above)
Comment (tugrulates) [reply tools/autograd/derivatives.yaml]: It doesn't do a cast if a `dtype` has not been passed to the original forward call:

https://github.com/pytorch/pytorch/blob/517c7c98610402e2746586c78987c64c28e024aa/aten/src/ATen/templates/TensorMethods.h#L12-L16

File: aten/src/ATen/native/ReduceOps.cpp (hunk 5 above)
Comment (tugrulates) [reply aten/src/ATen/native/ReduceOps.cpp]: There was an attempt at some point but it seems to have been broken. Let me reorder by primary op.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 6 above)
Comment (tugrulates) [reply aten/src/ATen/native/ReduceOps.cpp]: Because the signatures are different:

- `prod(Tensor self, int64_t dim, ...`
- `sum(Tensor self, IntList[1] dim, ...`

One of the sum native impl can be re-used by othes but prod wasn't so lucky.

Let me take a stab at changing the prod signature to use `IntList[1]`. This should be backwards compatible.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (wanchaol) [reply aten/src/ATen/native/native_functions.yaml]: sounds good, as long as it only expose from nn.functional I think would be fine :)

File: aten/src/ATen/native/ReduceOps.cpp (hunk 6 above)
Comment (tugrulates) [reply aten/src/ATen/native/ReduceOps.cpp]: Actually, the current organization makes sense in its own way. `cumsum` and `cumprod` are not really reduce functions, they are cumulative.

I'll still change to order by function, I already spent way too much mental effort for grasping what's going on here.

Comment (tugrulates) [reply aten/src/ATen/native/ReduceOps.cpp]: > Let me take a stab at changing the prod signature to use IntList[1]. This should be backwards compatible.

Punting on this because `prod_backward` needs to be rewritten to support multi-dim. This needs to be a separate PR.

File: test/test_autograd.py (hunk 1 above)
Comment (gchanan) [reply test/test_autograd.py]: it's the addition of the None node that causes you to need retain_graph?

File: test/test_jit.py (hunk 10 above)
Comment (gchanan) [reply test/test_jit.py]: No, I don't see how test_onnx_export_script_module_if actually tests anything anymore.  It doesn't instantiate a ModelToExport or run it through onnx, etc.

File: test/test_jit.py (hunk 9 above)
Comment (gchanan) [reply test/test_jit.py]: test_onnx_export_script_module_overload_fail and test_onnx_export_script_module_unsupported_optional are character-for-character identical.

File: tools/autograd/derivatives.yaml (hunk 11 above)
Comment (gchanan) [reply tools/autograd/derivatives.yaml]: Sorry, by generate an additional cast I meant "trace an additional cast".

File: tools/autograd/gen_python_functions.py (hunk 7 above)
Comment (gchanan) [reply tools/autograd/gen_python_functions.py]: I see; not a fan but you don't have to change it :).

File: aten/src/ATen/native/ReduceOps.cpp (hunk 6 above)
Comment (gchanan) [reply aten/src/ATen/native/ReduceOps.cpp]: I would recommend ordering by what you think makes sense (with an explanation); I was just describing why the current ordering didn't make sense to me.

File: test/test_jit.py (hunk 9 above)
Comment (tugrulates) [reply test/test_jit.py]: Ah, lol. I wanted to test two different things and they ended up being the exact same code. I remember writing `with self.assertRaisesRegex` twice for this PR with some time in-between.

Will remove one of them.

File: test/test_jit.py (hunk 10 above)
Comment (tugrulates) [reply test/test_jit.py]: Ah, right sorry. Saw it now.

File: tools/autograd/derivatives.yaml (hunk 11 above)
Comment (tugrulates) [reply tools/autograd/derivatives.yaml]: Sorry for the noob questions. Do you mean JIT tracing the backwards path? In what circumstance will this happen? I've failed to create a test scenario so far.

Though, I assume it won't be traced if the sample input doesn't execute the copy path, which it won't.
```
>>> torch.tensor(1.0, requires_grad=True).max().to(torch.double).grad_fn
<CopyBackwards object at 0x7f695dd3f898>
>>> torch.tensor(1.0, requires_grad=True).max().to(torch.float).grad_fn
<MaxBackward1 object at 0x7f695dd3f860>
```

**Comment by tugrulates** [comment]
@gchanan This should be ready to go after fixing the conflicts. Let me know if there are more changes you'd like to see.

**Comment by tugrulates** [comment]
Giving up on this since.
